---
title: Preprocessing the embryo chimera data set
author:
- name: Aaron T. L. Lun
  affiliation: &CRUK Cancer Research UK Cambridge Institute, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom
- name: Jonathan A. Griffiths
  affiliation: *CRUK
date: "2019-02-26"
vignette: >
  %\VignetteIndexEntry{02. Embryo preprocessing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc_float: true
    titlecaps: false
bibliography: ref.bib
---

<!--
AL: asterisks below remove weird highlighting on my editor.
****
-->



# Introduction

We will demonstrate a comparative analysis using data from a study of the early mouse embryo [@pijuansala2019single].
E8.5 embryos were dissociated and used to generate scRNA-seq data from the 10X Genomics protocol [@zheng2017massively].
This was performed using chimeric embryos in which _Tal1_-knockout (KO) embryonic stem cells were injected into a wild-type (WT) blastocyst.
The aim is to study the function of _Tal1_ (a transcription factor implicated in haematopoietic differentiation) in embryonic development.

The experiment uses a paired design with two batches of samples.
Each batch was independently generated by pooling cells from 6-7 chimeric embryos,
fluorescence-activated cell sorting by the expression of the tdTomato marker to separate WT and KO cells,
and using the sorted cell partitions in the 10X protocol.
This resulted in four samples in total, with two (matched) replicates per genotype containing 10,000 cells per sample.

# Setting up the data

## Obtaining the counts

We obtain the count data through the *[BiocFileCache](https://bioconductor.org/packages/3.9/BiocFileCache)* framework.
This caches the data locally upon the first download, avoiding the need to repeat the download on subsequent analyses.


```r
library(BiocFileCache)
bfc <- BiocFileCache("raw_data", ask=FALSE)
count.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/raw_counts.mtx.gz"))
```

We load in the count data from the MatrixMarket format, using methods from the *[Matrix](https://CRAN.R-project.org/package=Matrix)* package:


```r
library(Matrix)
counts <- readMM(count.path)
dim(counts)
```

```
## [1] 29453 56122
```

For the purposes of this demonstration, we will downsample to a quarter of the cells in the original data set.
This is simply to stay within time and memory limits on the Bioconductor build system^[Namely, 1 hour with 8 GB RAM.], and can be skipped by users with more computing resources.


```r
# Taking only every 4th cell.
down.keep <- rep(c(FALSE, FALSE, FALSE, TRUE), length.out=ncol(counts))
counts <- counts[,down.keep,drop=FALSE]
```

Finally, we will convert the matrix into the `dgCMatrix` format, which is more space and time-efficient than the default `dgTMatrix` produced by `readMM()`.


```r
counts <- as(counts, "dgCMatrix") 
```



## Constructing a `SingleCellExperiment`

We download the cell- and gene-level metadata using *[BiocFileCache](https://bioconductor.org/packages/3.9/BiocFileCache)*, and read them into R.


```r
meta.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/meta.tab.gz"))
meta.tab <- read.delim(meta.path, stringsAsFactors=FALSE)
head(meta.tab)
```

```
##     cell          barcode sample stage tomato stage.mapped
## 1 cell_1 AAACCTGAGACTTGAA      1  E8.5   TRUE         E8.5
## 2 cell_2 AAACCTGAGAGTAATC      1  E8.5   TRUE        E7.75
## 3 cell_3 AAACCTGAGGCTCTTA      1  E8.5   TRUE        E8.25
## 4 cell_4 AAACCTGAGGTAAACT      1  E8.5   TRUE        E8.25
## 5 cell_5 AAACCTGAGTGGTCCC      1  E8.5   TRUE         E8.0
## 6 cell_6 AAACCTGCAAAGAATC      1  E8.5   TRUE         E8.5
##                celltype.mapped haem_closestcell haem_subcluster
## 1 Forebrain/Midbrain/Hindbrain             <NA>              NA
## 2         Rostral neurectoderm             <NA>              NA
## 3            Paraxial mesoderm             <NA>              NA
## 4                 ExE mesoderm             <NA>              NA
## 5            Paraxial mesoderm             <NA>              NA
## 6               Cardiomyocytes             <NA>              NA
```

```r
gene.path <- bfcrpath(bfc, file.path("https://content.cruk.cam.ac.uk/",
    "jmlab/chimera_tal1_data/genes.tsv.gz"))
gene.tab <- read.delim(gene.path, header=FALSE, stringsAsFactors=FALSE)
colnames(gene.tab) <- c("ENSEMBL", "SYMBOL")
head(gene.tab)
```

```
##              ENSEMBL  SYMBOL
## 1 ENSMUSG00000051951    Xkr4
## 2 ENSMUSG00000089699  Gm1992
## 3 ENSMUSG00000102343 Gm37381
## 4 ENSMUSG00000025900     Rp1
## 5 ENSMUSG00000025902   Sox17
## 6 ENSMUSG00000104328 Gm37323
```

We store the count matrix in a `SingleCellExperiment` object with the metadata associated with each cell and gene.


```r
library(SingleCellExperiment)
sce <- SingleCellExperiment(list(counts=counts), 
    colData=meta.tab[down.keep,], rowData=gene.tab)
sce
```

```
## class: SingleCellExperiment 
## dim: 29453 14030 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(2): ENSEMBL SYMBOL
## colnames: NULL
## colData names(9): cell barcode ... haem_closestcell haem_subcluster
## reducedDimNames(0):
## spikeNames(0):
```

This object contains cells from all four samples.
KO cells should stain positive for the tdTomato marker.
Samples 1 and 3 come from the first batch, while samples 2 and 4 come from the second batch.


```r
sce$sample <- factor(sce$sample)
table(sce$sample, sce$tomato)
```

```
##    
##     FALSE TRUE
##   1     0 3687
##   2     0 3389
##   3  3380    0
##   4  3574    0
```



We set the row names to the gene symbols, augmented with the Ensembl identifier if the symbol is not unique or missing.


```r
library(scater)
rownames(sce) <- uniquifyFeatureNames(rowData(sce)$ENSEMBL, rowData(sce)$SYMBOL)
head(rownames(sce))
```

```
## [1] "Xkr4"    "Gm1992"  "Gm37381" "Rp1"     "Sox17"   "Gm37323"
```

# Quality control on the cells

Quality control (QC) on this data set has already been performed using a procedure similar to that described [here](https://bioconductor.org/packages/3.9/simpleSingleCell/vignettes/tenx.html).

- Cell-containing droplets were identified using the `emptyDrops()` function from the *[DropletUtils](https://bioconductor.org/packages/3.9/DropletUtils)* package at a false discovery rate of 1%.
See [here](https://bioconductor.org/packages/3.9/simpleSingleCell/vignettes/tenx.html#calling-cells-from-empty-droplets) for more details.
- Cells with high mitochondrial content were removed using an outlier-based approach as discussed [here](https://bioconductor.org/packages/3.9/simpleSingleCell/vignettes/tenx.html#quality-control-on-the-cells).
This removes putative damaged cells that have lost much of their cytoplasmic mRNA.

These steps can be performed separately per sample so existing QC methods for 10X data can be directly applied.
At this point, there is no need to explicitly use information across samples.
However, users should continue to inspect diagnostic statistics like the percentage of cells discarded in each sample.
Large differences between samples may be indicative of (i) strong biological differences at best or 
(ii) underlying technical problems that could compromise the downstream comparative analysis.

The code used for the processing of the embryo data set is available at https://github.com/MarioniLab/EmbryoTimecourse2018.
Note that, in this case, the mitochondrial filtering was performed on the pool of cells from all samples.
This is not strictly necessary.

# Normalization of cell-specific biases

We will use the `computeSumFactors()` function from *[scran](https://bioconductor.org/packages/3.9/scran)* to calculate size factors for each cell.
This removes scaling biases associated with cell-specific differences in capture efficiency, sequencing depth and composition biases. 
We refer users to the [previous workflow](https://bioconductor.org/packages/3.9/simpleSingleCell/vignettes/tenx.html#normalizing-for-cell-specific-biases) for more details on the method.
Note that we set `block=` to perform pre-clustering within each sample to reduce computational work.


```r
library(scran)

set.seed(100) # For the irlba usage.
clusters <- quickCluster(sce, use.ranks=FALSE, pc.approx=TRUE, block=sce$sample)
length(unique(clusters))
```

```
## [1] 57
```

```r
sce <- computeSumFactors(sce, clusters=clusters, min.mean=0.1)
summary(sizeFactors(sce))
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.2542  0.6855  0.8618  1.0000  1.1618  5.3250
```

We check that the size factors are roughly aligned with the total library sizes (Figure \@ref(fig:normplot)).
Deviations from the diagonal correspond to composition biases due to differential expression between cell subpopulations.
In this case, the cells below the diagonal probably correspond to erythrocytes that are dominated by hemoglobin upregulation.


```r
plot(Matrix::colSums(counts(sce)), sizeFactors(sce), xlab="Total library size", 
    ylab="Size factor", log="xy")
```

<div class="figure">
<img src="embryo_preprocess_files/figure-html/normplot-1.png" alt="Size factor estimates for each cell, plotted against the total library size." width="100%" />
<p class="caption">(\#fig:normplot)Size factor estimates for each cell, plotted against the total library size.</p>
</div>

We use the size factors to compute log-transformed normalized expression values, stored as the `"logcounts"` assay.


```r
sce <- normalize(sce)
assayNames(sce)
```

```
## [1] "counts"    "logcounts"
```

In practice, it is possible to perform this step separately for each sample, and then to combine the resulting `SingleCellExperiment` objects together.
This may be logistically simpler but requires a call to `multiBatchNorm()` to standardize the size factors to the same scale in the combined object.
See [here](https://bioconductor.org/packages/3.9/simpleSingleCell/vignettes/batch.html#feature-selection-across-batches) for a demonstration of how `multiBatchNorm()` might be used.

**Comments from Aaron:**

- Both `quickCluster()` and `computeSumFactors()` can be made considerably faster by turning on parallelization via the `block.BPPARAM=` and `BPPARAM=` arguments, respectively.
This typically comes at the cost of increased memory usage.

# Concluding remarks

Once the initial processing is done, we can save the resulting `SingleCellExperiment` object to file.
This will be used when merging the different samples onto the same coordinate system.


```r
saveRDS(sce, "embryo_processed.rds")
```

All software packages used in this workflow are publicly available from the Comprehensive R Archive Network (https://cran.r-project.org) or the Bioconductor project (http://bioconductor.org). 
The specific version numbers of the packages used are shown below, along with the version of the R installation.


```r
sessionInfo()
```

```
## R Under development (unstable) (2019-02-19 r76128)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.5 LTS
## 
## Matrix products: default
## BLAS: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRblas.so
## LAPACK: /home/cri.camres.org/lun01/Software/R/trunk/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scran_1.11.20               scater_1.11.10             
##  [3] ggplot2_3.1.0               SingleCellExperiment_1.5.2 
##  [5] SummarizedExperiment_1.13.0 DelayedArray_0.9.8         
##  [7] BiocParallel_1.17.15        matrixStats_0.54.0         
##  [9] Biobase_2.43.1              GenomicRanges_1.35.1       
## [11] GenomeInfoDb_1.19.2         IRanges_2.17.4             
## [13] S4Vectors_0.21.10           BiocGenerics_0.29.1        
## [15] Matrix_1.2-16               BiocFileCache_1.7.0        
## [17] dbplyr_1.3.0                BiocStyle_2.11.0           
## 
## loaded via a namespace (and not attached):
##  [1] viridis_0.5.1            httr_1.4.0              
##  [3] dynamicTreeCut_1.63-1    edgeR_3.25.3            
##  [5] bit64_0.9-7              viridisLite_0.3.0       
##  [7] DelayedMatrixStats_1.5.2 assertthat_0.2.0        
##  [9] statmod_1.4.30           BiocManager_1.30.4      
## [11] blob_1.1.1               GenomeInfoDbData_1.2.0  
## [13] vipor_0.4.5              yaml_2.2.0              
## [15] pillar_1.3.1             RSQLite_2.1.1           
## [17] lattice_0.20-38          glue_1.3.0              
## [19] limma_3.39.12            digest_0.6.18           
## [21] XVector_0.23.0           colorspace_1.4-0        
## [23] htmltools_0.3.6          plyr_1.8.4              
## [25] pkgconfig_2.0.2          bookdown_0.9            
## [27] zlibbioc_1.29.0          purrr_0.3.0             
## [29] scales_1.0.0             tibble_2.0.1            
## [31] withr_2.1.2              lazyeval_0.2.1          
## [33] magrittr_1.5             crayon_1.3.4            
## [35] memoise_1.1.0            evaluate_0.13           
## [37] beeswarm_0.2.3           tools_3.6.0             
## [39] stringr_1.4.0            locfit_1.5-9.1          
## [41] munsell_0.5.0            irlba_2.3.3             
## [43] compiler_3.6.0           rlang_0.3.1             
## [45] grid_3.6.0               RCurl_1.95-4.11         
## [47] BiocNeighbors_1.1.11     rappdirs_0.3.1          
## [49] igraph_1.2.4             bitops_1.0-6            
## [51] rmarkdown_1.11           codetools_0.2-16        
## [53] gtable_0.2.0             DBI_1.0.0               
## [55] curl_3.3                 R6_2.4.0                
## [57] gridExtra_2.3            knitr_1.21              
## [59] dplyr_0.8.0.1            bit_1.1-14              
## [61] stringi_1.3.1            ggbeeswarm_0.6.0        
## [63] Rcpp_1.0.0               tidyselect_0.2.5        
## [65] xfun_0.5
```

# References
